### PR TITLE
Porting pd and subcommands to newshell

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- [x] `pd` disassemble N instructions (also backward)
- [ ] `pd--` context disassembly of N instructions
- [x] `pda` disassemble all possible opcodes (byte per byte)
- [x] `pdb` disassemble basic block
- [x] `pdC` show comments found in N instructions
- [x] `pde` disassemble N instructions following execution flow from current PC
    - low effort on this since is executing esil & mapped hidden modes.
- [x] `pdf` disassemble function
    - also mapped hidden mode `summary`.
- [x] `pdi` like 'pi', with offset and bytes
    - moved as `pdq` since it is just a variation of `pd`
- [x] `pdj` disassemble to json
- [x] `pdJ` formatted disassembly like pd as json
- [x] `pdk` disassemble all methods of a class
- [x] `pdl` show instruction sizes
    - added also json version
- [x] `pdp` disassemble by following pointers to read ropchains
    - added also json version & fixed execution.
- [x] `pdr` recursive disassemble across the function graph
- [x] `pdR` recursive disassemble block size bytes without analyzing functions
- [x] `pdr.` recursive disassemble across the function graph (from current basic block)
- [x] `pds` disassemble summary (strings, calls, jumps, refs) (see pdsf and pdfs)
- [x] `pdt` disassemble N instructions in a table (see dtd for debug traces)
    - under pd and fixed output since it was completely broken.
- [x] `pdx` alias for pad or pix
    - removed since it is an alias for other 2 cmds

### CODE REVIEW

File: librz/core/cmd/cmd_print.c [hunk 1]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	}
+	rz_cmd_state_output_array_end(state);
+
+	free(block);
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_disassembly_basic_block_as_text_json_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	RzAnalysisBlock *b = rz_analysis_find_most_relevant_block_in(core->analysis, core->offset);
+	core->num->value = 0;
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7451]: shouldn't this line be moved *after* the check?

File: librz/core/cmd/cmd_print.c [hunk 2]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	}
+
+	ut8 *block = malloc(b->size + 1);
+	if (!block) {
+		RZ_LOG_ERROR("Cannot allocate buffer\n");
+		return RZ_CMD_STATUS_ERROR;
+	}
+	rz_io_read_at(core->io, b->addr, block, b->size);
+
+	rz_core_print_disasm(core->print, core, b->addr, block, b->size, 9999, 0, 2, true, state->d.pj, NULL);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c]: What about passing the whole `state` to the `rz_core_print_disasm()` instead?

File: librz/core/cmd/cmd_print.c [hunk 3]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	if (!function) {
+		function = rz_analysis_get_fcn_in(core->analysis, core->offset, 0);
+	}
+	if (!function) {
+		RZ_LOG_ERROR("Cannot find function at 0x%08" PFMT64x "\n", core->offset);
+		return RZ_CMD_STATUS_ERROR;
+	}
+
+	if (state->mode == RZ_OUTPUT_MODE_JSON) {
+		bool ret = rz_core_print_function_disasm_json(core, function, state->d.pj);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7526]: Same here - maybe remove `rz_core_print_function_disasm_json()` and pass the whole state to `rz_core_print_function_disasm()` and call that JSON code when available?

File: librz/core/cmd/cmd_print.c [hunk 4]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+		core->analysis->stackptr = saved_stackptr;
+		rz_config_set_i(core->config, "asm.lines.bb", asm_lines);
+	}
+	rz_config_set(core->config, "asm.bb.middle", orig_bb_middle);
+	return true;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_disassemble_recursively_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	RzAnalysisFunction *function = rz_analysis_get_fcn_in(core->analysis, core->offset, 0);
+	// RZ_ANALYSIS_FCN_TYPE_FCN|RZ_ANALYSIS_FCN_TYPE_SYM);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7862]: I think it is worth expanding this comment a bit.

File: librz/core/cmd/cmd_print.c [hunk 5]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+
+	rz_cmd_state_output_array_start(state);
+	bool ret = core_walk_function_blocks(core, function, state, 'D', false);
+	rz_cmd_state_output_array_end(state);
+	return ret ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_disassemble_recursively_from_current_block_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	RzAnalysisFunction *function = rz_analysis_get_fcn_in(core->analysis, core->offset, 0);
+	// RZ_ANALYSIS_FCN_TYPE_FCN|RZ_ANALYSIS_FCN_TYPE_SYM);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7876]: I think it is worth expanding this comment a bit

**Review by XVilka — CHANGES_REQUESTED** [review]
Amazing work! `p` commands are the trickiest ones.

File: librz/core/cmd/cmd_print.c [hunk 6]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_STANDARD:
+		rz_core_print_disasm(core->print, core, core->offset, core->block, core->blocksize, RZ_ABS(n_instrs), 0, 1, false, NULL, NULL);
+		break;
+	case RZ_OUTPUT_MODE_TABLE:
+		disassembly_n_instructions_as_table(state->d.t, core, RZ_ABS(n_instrs));
+		break;
+	case RZ_OUTPUT_MODE_JSON:
+		rz_cmd_state_output_array_start(state);
+		rz_core_print_disasm_json(core, core->offset, core->block, core->blocksize, RZ_ABS(n_instrs), state->d.pj);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7254]: It's a bit strange having the different parameters for the `rz_core_print_disasm_json()` with `rz_core_print_disasm()`. I think it should be one function, at the very least dispatching the calls, and accept `RzCmdStateOutput` instead.

File: librz/core/cmd/cmd_print.c [hunk 7]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	if (old_blocksize < n_bytes) {
+		rz_core_block_size(core, old_blocksize);
+		rz_core_block_read(core);
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_disassembly_basic_block_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	RzAnalysisBlock *b = rz_analysis_find_most_relevant_block_in(core->analysis, core->offset);
+	core->num->value = 0;
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7417]: Please move this line after the check, if possible

File: librz/core/cmd/cmd_print.c [hunk 8]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	}
+	rz_io_read_at(core->io, b->addr, block, b->size);
+
+	rz_cmd_state_output_array_start(state);
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_STANDARD:
+		rz_core_print_disasm(core->print, core, b->addr, block, b->size, 9999, 0, 2, false, NULL, NULL);
+		break;
+	case RZ_OUTPUT_MODE_JSON:
+		rz_core_print_disasm_json(core, b->addr, block, b->size, 0, state->d.pj);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7437]: The same concern applies here - shouldn't we pass `state` instead and merge these two functions?

File: librz/core/cmd/cmd_print.c [hunk 9]
```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
...
+	rz_list_sort(f->bbs, (RzListComparator)bbcmp);
+	if (state->mode == RZ_OUTPUT_MODE_JSON) {
+		rz_list_foreach (f->bbs, iter, b) {
+			ut8 *buf = malloc(b->size);
+			if (!buf) {
+				RZ_LOG_ERROR("cannot allocate %" PFMT64u " byte(s)\n", b->size);
+				return false;
+			}
+			(void)rz_io_read_at(core->io, b->addr, buf, b->size);
+			rz_core_print_disasm_json(core, b->addr, buf, b->size, 0, state->d.pj);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:7829]: Same concern applies here

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd/cmd_print.c (hunk 4 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:7862]: these comes from the original code. i do not know why they are like this, so i kept the comment. so much stuff needs to be rewritten

File: librz/core/cmd/cmd_print.c (hunk 1 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:7451]: it returned 0 on error in the old code.

File: librz/core/cmd/cmd_print.c (hunk 2 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c]: the whole code needs to be refactored and i'm not doing this in the PR

File: librz/core/cmd/cmd_print.c (hunk 6 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:7254]: both functions needs to be rewritten. it's a big mess and i'm not going to change them in this PR.

**Comment by wargio** [comment]
i did not port `pd--` because in my humble opinion it should be removed.

File: librz/core/cmd/cmd_print.c (hunk 6 above)
Comment (XVilka) [reply librz/core/cmd/cmd_print.c:7254]: Maybe we should open an issue to track this refactor.

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@imbillow please note this change since it might affect your work.

File: librz/core/cmd/cmd_print.c (hunk 6 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:7254]: https://github.com/rizinorg/rizin/issues/2554

**Pull request merged**